# Python과 Jupyter의 기초

![](https://i.imgur.com/gvSnw4A.png)

### Part 1 of "Python을 사용한 데이터 분석: 0부터 Pandas까지"

본 튜토리얼 시리즈는 초보자를 위한 Python을 이용한 프로그래밍과 데이터 분석을 소개합니다.   
이 튜토리얼은 실용적이고 코딩 중심적인 접근으로 진행됩니다.
본 튜토리얼을 효율적으로 학습하는 가장 좋은 방법은 코드를 실행하고 직접 사용해 보는 것입니다.


1. [Python과 Jupyter 기초]
2. [변수 및 데이터 유형 요약 알아보기]
3. [조건문 및 루프를 사용한 분기]
4. [함수를 이용한 재사용 가능한 코드 작성]
5. [파일 읽기 및 쓰기]
6. [Python과 Numpy를 이용한 수치형 연산]
7. [Pandas를 이용한 데이터 표 분석]
8. [Matplotlib와 Seaborn을 사용한 데이터 시각화]
9. [탐색적 데이터 분석 - 사례 연구]

해당 튜토리얼에서는 다음 주제를 다룹니다.:

* Python을 사용하여 산술 연산 수행
* 변수를 사용한 multi-step 문제 해결
* Python을 사용한 조건의 참/거짓 판단
* 조건과 논리연산자의 결합
* Markdown을 사용한 텍스트 스타일 추가


### 코드 실행 방법

### Option 1: 무료 온라인 리소스를 사용하여 실행

무료로 제공되는 온라인 리소스를 사용하여 실행하는 방법입니다.  
**Run** 버튼을 클릭한 뒤 **Run on Binder**를 선택하여 해당코드를 실행시킬수 있습니다.  

"Run on Colab" 혹은 "Run on Kaggle"을 선택하여 [Google Colab](https://colab.research.google.com) 혹은[Kaggle](https://kaggle.com) 이 플랫폼에서도 해당코드를 실행시킬 수 있지만,  
이러한 플랫폼을 사용하려면 [Google Colab](https://colab.research.google.com) 또는 [Kaggle](https://kaggle.com)에서 계정을 생성해야 합니다.

<!--#### Option 2: Running on your computer locally-->
### Option 2: 컴퓨터에서 로컬로 실행
로컬 환경에서 해당코드를 실행시키려면 [Python](https://www.python.org)이 필요하고, Jupyter Notebooks을 다운로드 한 후 필요한 라이브러리를 설치해야 합니다.  
Python 환경을 분리시켜 관리해주는 [Conda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/)를 활용하여 환경설정을 하는 것을 추천합니다.


> **Jypyter Notebooks** : 이 튜토리얼은 [Jupyter notebook](https://jupyter.org)으로 만들었습니다.
각 셀은 Python으로 작성된 코드나 간단한 설명을 포함할 수 있습니다.
코드 셀을 실행하고, 실행결과(숫자, 문자, 그래프, 테이블, 파일 등)를 Jupyter Notebook 내에서 즉시 볼 수 있습니다.
Jupyter는 실험과 분석을 위한 플랫폼이므로, 코드가 지저분하게 작성되어있거나 여러 셀에 나뉘어진 상태에서 문제를 해결하는 것을 절대 두려워하지 마십시오.
오류를 발견하고 수정하는 과정에서 많은 것을 배울 수 있으며, 상단의 "모든 셀의 출력 지우기" 옵션을 사용하여 모든 출력을 지우고 위에서부터 다시 시작할 수 있습니다.


## <span style='color:black; background-color:#fff5b1;'>Python을 사용한 산술 연산 수행</span>

먼저, Python을 계산기로 사용해보겠습니다. Jupyter 내에서 코드 셀을 사용하여 Python을 작성하고 실행할 수 있습니다.

> **셀 작업** : Jupyter 내에 새로운 셀을 생성하려면 메뉴 모음에서 "삽입 > 아래에 셀 삽입"을 선택하거나, 도구 모음에서 "+" 버튼을 누르면 됩니다.  
또한 단축키 Esc+B를 사용하여 새로운 셀을 생성할 수 있습니다.   
셀이 만들어지면 셀을 클릭하여 선택하고, 그런 다음 "셀 > 셀 유형" 메뉴 선택사항을 사용하여 셀 유형을 코드 또는 Markdown(텍스트)로 변경할 수 있습니다.  
또한 단축키인 Esc+Y와 Esc+M을 사용할 수 있으며, 셀을 두 번 클릭하여 셀 내의 내용을 편집할 수 있습니다.  
변경사항을 적용하고 셀을 실행하려면 "셀 > 셀 실행" 메뉴 옵션을 사용하거나, 도구 모음에서 "실행" 단추를 클릭하거나 "Shift+Enter"를 사용하면 됩니다.


아래의 코드 셀을 실행하여 계산을 수행하고, 결과를 확인하세요.  
그런 다음, 본인의 생각에 따라 숫자를 변경하고 수정된 셀을 다시 실행하여 결과를 확인하세요.  
'//', '%', '**' 연산자가 무엇에 쓰이는지 짐작할 수 있나요?





In [2]:
2 + 3 + 9

14

In [6]:
99 - 73

26

In [7]:
23.54 * -1432

-33709.28

In [8]:
100 / 7

14.285714285714286

In [9]:
100 // 7

14

In [10]:
100 % 7

2

In [11]:
5 ** 3

125

예상할 수 있듯이, `/`, `*`와 같은 연산자는 수학적 관례에 따라 `+`, `-`와 같은 다른 연산자보다 우선순위가 높습니다.  
수학과 동일하게, 괄호 ()를 사용하여 작업 수행 순서를 지정할 수 있습니다.


In [ ]:
((2 + 5) * (17 - 3)) / (4 ** 3)

Python은 다음과 같은 산술 연산자를 지원합니다 :
| 연산자      | 목적               | 예시        | 결과      |
|------------|--------------------|-------------|-----------|
| `+`        | 덧셈               | `2 + 3`     | `5`       |
| `-`        | 뺄셈               | `3 - 2`     | `1`       |
| `*`        | 곱셈               | `8 * 12`    | `96`      |
| `/`        | 나눗셈             | `100 / 7`   | `14.28..` |
| `//`       | 몫                 | `100 // 7`  | `14`      |    
| `%`        | 나머지             | `100 % 7`   | `2`       |
| `**`       | 제곱               | `5 ** 3`    | `125`     |


### <span style='color:black; background-color:#f5f0ff;'>Python의 명명 규칙</span>
Python에서는 PEP 8 스타일 가이드를 따라 작성하는 것이 좋습니다.
>`PEP 8`: `PEP 8`이란 Python 코드를 어떤 형식으로 작성할지 알려주는 스타일 가이드로, Python 언어의 여러 부분에 사용하는 이름을 어떻게 붙일지에 대한 고유 스타일을 제공합니다.  
이런 규칙에 따라 작성하면, 코드를 읽을 때 각 이름이 어떤 유형에 속하는지 쉽게 구분할 수 있습니다.

- 함수, 변수, 애트리뷰트(attribute)는 lowercase_underscore처럼 소문자와 밑줄을 사용한다.
- 보호돼야 하는 인스턴스 애트리뷰트는 일반적인 애트리뷰트 이름 규칙을 따르되, _leading_underscore처럼 밑줄로 시작한다.
- 비공개(private)(한 클래스 안에서만 쓰이고 다른 곳에서는 쓰면 안되는 경우) 인스턴스 애트리뷰트는 일반적인 애트리뷰트 이름 규칙을 따르되, __leading_underscore처럼 밑줄 두 개로 시작한다.
- 클래스(예외도 포함한다)는 CapitalizedWord처럼 여러 단어를 이어 붙이되, 각 단어의 첫 글자를 대문자로 만든다.

## <span style='color:black; background-color:#fff5b1;'>변수를 사용한 multi-step 문제 해결</span>

Python을 사용하여 다음 문자 문제를 해결해보겠습니다.
> 한 식료품점은 얼음 한 봉지를 1.25달러에 팔고, 20%의 수익을 올린다. 만약 얼음 500봉지를 판매한다면, 총 이익은 얼마나 될까?

우리는 제공된 정보를 나열하고, 위 문제를 Python을 사용하여 수학적 표현으로 점진적으로 변환할 수 있습니다.  
*얼음 봉지 비용($)* = 1.25  

*이익률* = 20% = .2    

*한 봉지당 이익($)* = 이익률 * 얼음봉지 비용 = .2 *1.25  

*봉지 수* = 500  

*총 이익* = 봉지 수 * 봉지당 이익 = 
500 * (.2 *1.25)


In [ ]:
500 * (.2 * 1.25)

위 과정에 의해, 식료품점은 총 125달러의 수익을 올립니다. 이것이 문제를 해결하는 합리적인 방법이지만, 코드 셀을 보면 숫자가 의미하는 것이 무엇인지 명확히 알 수 없습니다.  
따라서 우리는 Python *변수*를 만들어 각각의 숫자에 이름을 붙일 것입니다.

> *변수(Variables)* : Python과 같은 프로그래밍 언어로 작업하는 동안 정보가 *변수*에 저장됩니다.  
변수는 데이터를 저장하기 위한 컨테이너라고 생각할 수 있으며, 변수 내에 저장된 데이터를 *값(value)* 라고 합니다.


In [13]:
cost_of_ice_bag = 1.25

In [14]:
profit_margin = .2

In [15]:
number_of_bags = 500

위 셀에서 볼 수 있듯이, cost_of_ice_bag, profit_margin, number_of_bags 변수에는 이제 문제에 제공된 정보가 포함됩니다.  
셀에 변수 이름을 입력하여 변수 값을 확인할 수 있으며, 산술연산을 사용하여 변수를 결합하는 등 다른 변수를 만들 수 있습니다.

> **코드 완성** : Jupyter 내의 코드 셀에 기존 변수의 이름을 입력할 때, 처음 몇 글자를 입력하고 `Tab`키를 누르면 변수 이름이 자동으로 완성됩니다.  
아래의 코드 셀에 'pro'를 입력하고 `Tab`을 눌러 'profit_margin'으로 자동완성 해보세요.


In [16]:
profit_margin

0.2

In [17]:
profit_per_bag = cost_of_ice_bag * profit_margin

In [18]:
profit_per_bag

0.25

In [19]:
total_profit = number_of_bags * profit_per_bag

In [20]:
total_profit

125.0

만약 *정의* 되지 않은 변수의 값을 보려고 하면, Python이 오류를 표시합니다.
<!--If you try to view the value of a variable that has not been *defined*, i.e., given a value using the assignment statement `variable_name = value`, Python shows an error.-->

In [21]:
net_profit

NameError: name 'net_profit' is not defined

적절한 이름의 변수를 사용하여 데이터를 저장하고 조작하는 것은 코드가 수행하는 작업을 설명할 수 있는 좋은 방법입니다.  
메시지를 사용하여 문제의 결과를 표시해봅시다. 'print'*함수(function)* 를 사용하여 이 작업을 수행할 수 있습니다.  
> **함수(Functions)** : 함수는 재사용 가능한 기능들의 집합입니다. 하나 이상의 입력을 사용하고, 특정 연산을 수행하며, 종종 출력을 반환합니다.  
Python은 'print'와 같은 많은 내장 함수를 제공하며, 자체적으로 함수를 정의할 수도 있습니다.  


In [ ]:
print("The grocery store makes a total profit of $", total_profit)

> **print** : `print` 함수는 정보를 출력하는 데 사용되며, 텍스트("안녕? 나는 텍스트야"), 숫자, 변수, 수학식 등 하나 이상의 입력을 필요로 합니다.  
다음 튜토리얼에서는 변수와 함수에 대해 자세히 알아보겠습니다.

각 변수 또는 수학적 연산에 대한 코드 셀을 만드는 것이 지루할 수 있지만, 다행히 Jupyter를 사용하면 단일 코드 셀 내에 여러 줄의 코드를 작성할 수 있습니다.
셀 내의 마지막 코드 라인의 결과가 출력으로 표시됩니다.

하나의 셀 안에 문제에 대한 해결책을 다시 작성해보자.


In [ ]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

코드 내에 *주석(comments)* 을 추가하기 위해 `#` 문자를 사용합니다.
> **주석(comments)** : 주석과 빈 줄은 실행에 영향을 끼치지 않지만, 코드가 수행하는 작업에 대한 정보를 사람에게 제공하는데 유용합니다.  
주석은 inline(일부 코드의 끝), 별도의 줄에 있거나, 여러 줄에 걸쳐있을 수 있습니다.

inline 및 단일 줄 주석은 `#` 로 시작하는 반면, 주석이 길어 여러 줄에 걸쳐있어야 할 경우 세 개의 따옴표로 시작하고 끝납니다. (""")

아래는 주석의 예시입니다.


In [ ]:
my_favorite_number = 1 # an inline comment

In [ ]:
# This comment gets its own line
my_least_favorite_number = 3

In [ ]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5

## <span style='color:black; background-color:#fff5b1;'>Python을 사용한 피연산자 평가</span>

Python은 산술 연산 외에도 숫자와 변수를 비교하기 위한 몇 가지 연산도 제공합니다.
| 연산자      | 설명                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | 피연산자가 동일한지 확인                                     |
| `!=`        | 피연산자가 동일하지 않은지 확인                                 |
| `>`         | 좌측 피연산자가 우측 피연산보다 큰지 확인             |
| `<`         | 좌측 피연산자가 우측 피연산자보다 작은지 확인                |
| `>=`        | 좌측 피연산자가 우측 피연산자보다 크거나 같은지 확인 |
| `<=`        | 우측 피연산자가 좌측 피연산자보다 크거나 같은지 확인    |

비교 연산의 결과는 `참(True)` 또는 `거짓(False)` 입니다.(T/F)  
이들은 Python의 특별한 키워드 입니다. 아래 셀에서 비교 연산자를 사용하여 몇 가지 실험을 진행해보겠습니다.


In [ ]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [ ]:
# Equality check - True
my_favorite_number == 1

In [ ]:
# Equality check - False
my_favorite_number == my_least_favorite_number

In [ ]:
# Not equal check - True
my_favorite_number != a_neutral_number

In [ ]:
# Not equal check - False
a_neutral_number != 3

In [ ]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

In [ ]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

In [ ]:
# Less than check - True
my_favorite_number < 10

In [ ]:
# Less than check - False
my_least_favorite_number < my_favorite_number

In [ ]:
# Greater than or equal check - True
my_favorite_number >= 1

In [ ]:
# Greater than or equal check - False
my_favorite_number >= 3

In [ ]:
# Less than or equal check - True
3 + 6 <= 9

In [ ]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

산술 연산과 마찬가지로 비교 연산 결과도 변수에 저장할 수 있습니다.


In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

## <span style='color:black; background-color:#fff5b1;'>피연산자를 논리 연산자와 결합</span>
논리 연산자인 `and`, `or`, `not`은 조건 및 `True`와 `False`값 (T/F값을 booleans 라고도 함)에 따라 동작합니다.  
`and`와 `or`은 두 개의 피연산자에서 동작하는 반면, `not`은 한 개의 피연산자에서 동작합니다.

`and` 연산자는 두 개의 피연산자가 모두 `True`로 평가될 때, `True`를 반환합니다. 그렇지 않으면 `False`가 반환됩니다.


In [ ]:
my_favorite_number

In [ ]:
my_favorite_number > 0 and my_favorite_number <= 3

In [ ]:
my_favorite_number < 0 and my_favorite_number <= 3

In [ ]:
my_favorite_number > 0 and my_favorite_number >= 3

In [ ]:
True and False

In [ ]:
True and True

`or` 연산자는 하나 이상의 피연산자가 `True`로 평가될 경우 `True`를 반환하고, 두 피연산자가 모두 `False`일 경우에만 `False`를 반환합니다.


In [ ]:
a_neutral_number = 3

In [ ]:
a_neutral_number == 3 or my_favorite_number < 0

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

In [ ]:
my_favorite_number < 0 or True

In [ ]:
False or False

`not` 연산자는 피연산자가 `True`이면 `False`를 반환하고, 피연산자가 `False`일 경우 `True`를 반환합니다.


In [ ]:
not a_neutral_number == 3

In [ ]:
not my_favorite_number < 0

In [ ]:
not False

In [ ]:
not True

논리 연산자는 복잡한 연산식을 형성하기 위해 결합될 수 있습니다.  
논리 연산자를 적용해야 하는 순서를 명확히 나타내려면, 괄호 ()를 사용합니다.


In [ ]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

In [ ]:
not (True and 0 < 1) or (False and True)

괄호를 사용하지 않으면 논리 연산자의 계산 순서는 왼쪽에서 오른쪽으로 진행됩니다.
<!--If parentheses are not used, logical operators are applied from left to right.-->

In [ ]:
not True and 0 < 1 or False and True

JupyterNotebook의 대화영 특성을 사용하여, Python의 산술, 조건부 및 논리 연산를 실험합니다.  
이후 튜토리얼에서 변수와 함수에 대해 더 깊이 배우도록 하겠습니다.  


## <span style='color:black; background-color:#fff5b1;'>Markdown을 사용한 텍스트 스타일 추가</span>

텍스트 셀을 사용하여 설명을 추가하는 것은 본인을 포함한 다른 사용자에게 코드에 대한 유용한 정보를 제공하는 좋은 방법이며, 나중에 다시 참조해야 할 경우에도 유용하게 작용합니다.  
Jupyter 내에서 텍스트 셀을 두 번 눌러 편집할 수 있으며, 편집 모드에서는 텍스트가 약간 다르게 나타납니다.(ex. 머리 글에 ## 접두사가 있음)  
이 글 또한 텍스트에 스타일을 추가하는 간단한 방법인 Markdown을 사용하여 작성됩니다.  
`##`와 같은 특수문자 없이 출력을 보기 위해서는, 이 셀을 실행하면 됩니다.  

소스와 출력을 전환하며 Markdown을 이용한 특정 스타일을 적용해나갈 수 있습니다.  

예를 들어, 행의 시작 부분에 하나 이상의 `#` 문자를 사용하여 크기가 다른 헤더를 만들 수 있습니다.  
# 머리글 1
## 머리글 2
### 머리글 3
#### 머리글 4

글머리 기호 또는 번호 목록을 만드려면, `*` 또는 `1.x` 로 행을 시작합니다.  
글머리 기호 목록 :
* 내용 1
* 내용 2
* 내용 3

번호 목록 :
1. 사과
2. 바나나
3. 파인애플

또한 굵은 글씨를 사용하기 위해서는, `**` 를 양 옆에 사용하면 됩니다.  (' ** 텍스트 ** ')  
예를 들어, '**안녕 나는 굵은 글씨야**' 처럼요.  
  
혹은, 글씨에 기울임 효과를 적용하고 싶다면 `*`를 양 옆에 사용하면 됩니다.
('* 텍스트 *')  
예를 들어, *안녕 나는 기울어졌어* 처럼요.

Markdown의 또 다른 좋은 특징은, 코드 블록을 포함하는 기능입니다.  
**Markdown 셀 내부의 코드 블록은 실행할 수 없습니다.**

### <span style='color:black; background-color:#dcffe4;'>필요한 계산 수행</span>
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

### <span style='color:black; background-color:#dcffe4;'>결과 표시</span>
print("The grocery store makes a total profit of $", total_profit)

